docs: Restructures developer section

vikram-dagger opened the pull request:
👀 [Preview](https://deploy-preview-6681--devel-docs-dagger-io.netlify.app/zenith/developer-guide)

vikram-dagger commented:
@helderco @vito @TomChv can I get your feedback on the new structure, which is currently outlined only for the Python section? Based on feedback I will make the same changes for the Go and TypeScript sections. See https://deploy-preview-6681--devel-docs-dagger-io.netlify.app/zenith/developers/

vito commented:
@vikram-dagger I like this structure a lot more. Whenever I went to the Zenith docs the amount of sections beneath e.g. "Go" felt deceptively narrow:

![image](https://github.com/dagger/dagger/assets/1880/2d84d505-c2f6-4684-a2f0-8c2533439ae7)

This new structure feels much more discoverable, since there are many independent reasons someone might be going to the docs:

![image](https://github.com/dagger/dagger/assets/1880/2234c853-2cc7-45cb-af27-1965a36919d0)

I'm not 100% sure what I would expect to see under "Secrets" and "Services" though. I don't think there's a whole lot different between SDKs in that regard but I could be missing something. Those two might be better documented in a CLI-centric section, where we could also cover "Terminal".

helderco commented:
Same as @vito, this is much better 😀 

In Dependencies, there's actually quite a bit that can be said specifically to Python there, in terms of Python dependencies, not module dependencies.

vikram-dagger commented:
> @vikram-dagger I like this structure a lot more. Whenever I went to the Zenith docs the amount of sections beneath e.g. "Go" felt deceptively narrow:

Thanks for the feedback, glad this feels like a better structure.

> I'm not 100% sure what I would expect to see under "Secrets" and "Services" though. I don't think there's a whole lot different between SDKs in that regard but I could be missing something. Those two might be better documented in a `dagger call`-centric section, where we could also cover "Terminal".

My thinking was that those sections would have 1-2 code examples of passing / using secrets / services in a module.

helderco commented:
> > I'm not 100% sure what I would expect to see under "Secrets" and "Services" though. I don't think there's a whole lot different between SDKs in that regard but I could be missing something. Those two might be better documented in a `dagger call`-centric section, where we could also cover "Terminal".
> 
> My thinking was that those sections would have 1-2 code examples of passing / using secrets / services in a module.

Yeah, but since that's common between SDKs, it should be in multilingual. The SDK specific sections should be reserved for exactly that: SDK specific DX and considerations.

TomChv commented:
I love the new structure! It's way better :rocket:

vikram-dagger commented:
> Yeah, but since that's common between SDKs, it should be in multilingual. The SDK specific sections should be reserved for exactly that: SDK specific DX and considerations.

The plan atm is to first create language-specific sections for each SDK and then evaluate later to see which can be made common or multi-lingual. After some discussions, our first guess is that a Go developer would like to have a dedicated Go section with Go-only examples (and the same for Python and TS).

helderco commented:
> The plan atm is to first create language-specific sections for each SDK and then evaluate later to see which can be made common or multi-lingual. After some discussions, our first guess is that a Go developer would like to have a dedicated Go section with Go-only examples (and the same for Python and TS).

I think that approach leads to uncessary duplication. We already know what's common or not. Doesn't make sense for me to write a common thing like secrets in the Python SDK category, so we later evaluate if it should be multi-lingual or not.

I think what's better is to make sure we have **good multi-lingual docs as what you start with**, that cover secrets, including properly documenting fields for the API, and then use the SDK-specific categories as a more advanced or in-depth dive into the language specifics or details.

If we put the focus on a beginner starting from language specific docs rather than the common multilingual ones, it'll push us to duplicate and make language docs more complete on their own, thus leading to the duplication and defeating the better maintainability of multilingual examples.

vikram-dagger commented:
> I think what's better is to make sure we have **good multi-lingual docs as what you start with**, that cover secrets, including properly documenting fields for the API, and then use the SDK-specific categories as a more advanced or in-depth dive into the language specifics or details.

Could you please give an example of this, say for secrets? ie. what content do you feel should go into the common "fundamentals" section vs the SDK-specific sections?

helderco commented:
> Could you please give an example of this, say for secrets? ie. what content do you feel should go into the common "fundamentals" section vs the SDK-specific sections?

Using secrets works the same in all SDKs. It's just an argument of type `dagger.Secret`, where the value is usually passed from the host via the CLI. I think most of the docs should be multilingual as the SDKs are fairly consistent in what they support. 

The language specific sections can turn out to be pretty small, not about how to use Dagger, but more about how the SDK's DX integrates with the API.

helderco commented:
I've written a page today full of details on the Python project structure. It's not required reading, nor should it be a part of a tutorial on getting started. I think this sort of docs is useful when you're already hands-on and maybe are trying to do something not described elsewhere and looking for more details or explanations on how certain things are set up. 

So, more like a reference guide, not a how-to. It's actually the sort of thing I'd make a first pass read in the beginning just because I like to understand things, which usually helps me know what I can refer back to when I need to.

vikram-dagger commented:
A quick summary regarding the previous discussion on the developer docs structure, see  https://github.com/dagger/dagger/pull/6681#issuecomment-1952394278 and https://github.com/dagger/dagger/pull/6681#issuecomment-1952450472 for full details.

We will have to shortly decide between the two approaches mentioned:

A: A base set of common, multilingual docs + a set of docs for each language covering language-specific items
- advantage: easier to maintain for us as there is less duplication of content, smaller language-specific doc collections
- disadvantage: a dev needs to look through / jump between two sets of docs

B: An independent set of docs for each language
- advantage: a dev has everything needed for the selected language in one section of the docs
- disadvantage: duplication of content between languages, as there will be pages with similar content and only differing in code samples

(there may be other adv/disadv, feel free to add)

To help us see how these approaches could work in practice and arrive at a decision, I have created two parallel nav trees in the sidebar ([preview](https://deploy-preview-6681--devel-docs-dagger-io.netlify.app/zenith/developer-guide)), labelled as Option A and Option B. Please add your thoughts/comments on which "feels better" as an end-user of the docs.

helderco commented:
Reposting what I said in [discord](https://discord.com/channels/707636530424053791/1209230299389501491/1210012379727921192).

As I see it, the common docs and language docs serve different purposes. They're not a part of the same flow.

- **Common docs** *(practical)*
  - tutorial (learning oriented) or how-to guides (problem-oriented)
  - show by example
  - most common use cases that work across SDKs
  - short explanations
- **Language docs** *(theoretical)*
  - conceptual guides (understanding oriented) or reference guides (information oriented)
  - specific to SDK implementation details or DX

sipsma commented on docs/versioned_docs/version-zenith/developer-guide/go/637319-interfaces.mdx:
````diff
@@ -0,0 +1,43 @@
+}
+```
+
+<!-- TODO verify the following with Erik -->
````
Yep that's all correct!

vito commented on docs/versioned_docs/version-zenith/partials/_developer_troubleshooting.mdx:
just fyi, pushed a commit to switch this from `--focus=false` to `--debug`

vikram-dagger replied:
Thanks! @vito

sipsma commented on docs/versioned_docs/version-zenith/developer-guide/fundamentals/942201-execution-environment.mdx:
```diff
@@ -0,0 +1,43 @@
+
+As mentioned above, Functions execute in containers and thus do not have access to resources on the host you invoke the Function from (i.e. the host you execute a CLI command like `dagger call` from). Instead, these resources need to be explicitly passed when executing `dagger call`.
+
+This includes: (TODO:(sipsma) include code examples here? Or just link somewhere else?)
```
cc @vikram-dagger were you imagining these docs in `fundamentals` to include CLI examples too? Or should these be super high-level and just link to other sections? I'm happy to fill in other sections with any details mentioned here that are missing if so. Or can also just provide full examples here too. Just let me know what you think.

vikram-dagger replied:
Yes, of course, you can add CLI examples if you like. However for the specific items you have here, we already have various examples (see Arguments page for an example of passing a host/remote dir, Secrets page for passing an env var, Chaining and Quickstart pages for chaining examples, etc). So it's fine to link to those pages if easier. We don't yet have a Services page with an example, mainly due to time constraints so that link can be added later once a page is ready.

sipsma replied:
Thanks, updated to link to the "Pass argument to a function" guide here too!

vikram-dagger commented:
> I pushed a first draft of the architecture docs here: [f5aa505](https://github.com/dagger/dagger/commit/f5aa5053390fb327d19dc3b49e42f40895c214a4)
> 
> I ended up just re-using the existing "How does it work" section on the landing page almost entirely, only made some minor tweaks. It seemed like it pretty much filled the role of explaining the high level architecture, but let me know what you think @vikram-dagger and others.

Yes, this is pretty much exactly what I imagined we would do.

> I also updated the "How does it work" section on the landing page to just link to this section now to dedupe content.

Thanks!

vikram-dagger commented on docs/versioned_docs/version-zenith/developer-guide/fundamentals/483821-arguments-return-values.mdx:
```diff
@@ -178,6 +182,72 @@ The result will look like this:
+│   │           │   ├── _types.py
+│   │           │   ├── _utils.py
+│   │           │   └── cli.py
+│   │           └── py.typed
```
I've modified to one level only.

vikram-dagger commented on docs/versioned_docs/version-zenith/developer-guide/fundamentals/483821-arguments-return-values.mdx:
```diff
@@ -0,0 +1,495 @@
+.
+├── LICENSE
+├── dagger
+│   ├── dagger.gen.go
```
I've modified to one level only. I'm guessing this should be relatively stable (only shows the top-level 3 entries now) so not changing the example, but lmk if you still feel it should change.

helderco commented on docs/versioned_sidebars/version-zenith-sidebars.json:
```diff
@@ -192,15 +196,11 @@
-                {
-                  "type": "doc",
-                  "id": "developer-guide/python/visibility"
+                  "id": "developer-guide/python/ide-integration"
```
Breaking into fundamentals and further reading is one nesting too much inside the language docs. Especially since sdks have duplicated trees, they have the same tone.

vikram-dagger replied:
Done

helderco commented:
I've udpated the Python docs. I reordered the pages to a flow that makes sense, building from what came before.

vikram-dagger commented on docs/versioned_docs/version-zenith/developer-guide/python/snippets/first-module/step4/main.py:
```diff
@@ -0,0 +1,24 @@
+@function
+def hello_world(
+    count: Annotated[int, Doc("The number of potatoes to process")],
+    mashed: Annotated[bool, Doc("Whether the potatoes are mashed")] = False,
```
Done

vikram-dagger commented:
> I've udpated the Python docs. I reordered the pages to a flow that makes sense, building from what came before.

In https://github.com/dagger/dagger/pull/6681/commits/874c2d7fabc9284813a8d029c9618c01426c6e02 I've moved the dependencies page up in the sequence you suggested, because I think that is an important concept to discuss early. Apart from this, I updated the navbar for the other languages to match your suggestion.

helderco commented:
> > I've udpated the Python docs. I reordered the pages to a flow that makes sense, building from what came before.
> 
> In [874c2d7](https://github.com/dagger/dagger/commit/874c2d7fabc9284813a8d029c9618c01426c6e02) I've moved the dependencies page up in the sequence you suggested, because I think that is an important concept to discuss early. Apart from this, I updated the navbar for the other languages to match your suggestion.

Yeah, I agree. Also think that adding Python dependencies could be extracted into a page between module structure (I think at one point it was "Project structure", which is more accurate since it's talking mostly about a Python project managed with `pyprojec.toml`), and IDE integration. The latter was also a part of "project structure".

helderco commented on docs/versioned_docs/version-zenith/developer-guide/python/820256-module-structure.mdx:
````diff
@@ -135,27 +141,30 @@ from .lint import Lint  # in src/main/lint.py
 ```
 
-## src layout
+## Default project layout
````
"src layout" is a thing in Python (https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/), that's why I used it. The emphasis was intentional.

It used to be a common pitfall to not put sources in `src`, so there's lots writing on the web about using a "src layout", meaning, put all your python files in a `src` directory, that is **not your working directory**.

Now it's a convention that many tools default to.

vikram-dagger replied:
Didn't realize this convention, reverted

helderco commented on docs/versioned_docs/version-zenith/developer-guide/python/753720-arguments-return-values.mdx:
```diff
@@ -7,10 +7,12 @@ title: "Functions"
 import PartialExperimentalDocs from '../../partials/_experimental.mdx';
 
-# Functions
+# Functions, Arguments and Return Values
```
I changed this to just "Functions" on purpose. Arguments and return values **are a part** of any function in any language, including in dagger. The longer title is unnecessary in my opinion.

vikram-dagger replied:
Good point, changed in all

helderco commented on docs/versioned_docs/version-zenith/developer-guide/python/428321-constructors.mdx:
````diff
@@ -0,0 +1,108 @@
+generated `__init__()` function, using [dataclasses.field(init=False)][field]:
+
+```python file=./snippets/constructors/exclude/main.py
+```
````
Note that after "chaining", this brings the concept of the `dataclasses.field` descriptor, still without `dagger.field`.

helderco commented on docs/versioned_docs/version-zenith/developer-guide/typescript/817191-custom-types.mdx:
```diff
@@ -0,0 +1,34 @@
+
+TypeScript has multiple ways to support complex data types. However, the Dagger TypeScript SDK only supports it through fields (decorated with `@field()`) inside classes because `type` and `interface` keywords do not allows decorators.
+
+Here is an example of a function named `daggerOrganization` that returns a custom `GitHubOrganization` type, itself containing a collection of `GitHubAccount` types:
```
Users shouldn't prefix their custom types with the name of the module since the API already does that. I don't think we should encourage users to do it by seeing it in examples. In the Python version for this doc I removed the prefix for this reason. Otherwise you'll see in the API these types:
- GitHub
- GitHubGitHubOrganization
- GitHubGitHubAccount

helderco commented on docs/versioned_docs/version-zenith/developer-guide/typescript/snippets/custom-types/index.ts:
```diff
@@ -0,0 +1,48 @@
+}
+
+@object()
+class MyModule {
```
Since this seems to be a module reated to github I think the module name should be "github", making the main object "Github". Notice I didn't suggest "GitHub", since that would make the module name `git-hub` and constructor function `dag.gitHub()`.

Note that "custom types" are prefixed with the module's name in the API to avoid conflicts, so **we don't want users to add the prefix themselves**. In this case, current types will be:
- MyModule
- MyModuleGitHubAccount
- MyModuleGitHubOrganization

In the Python version for this doc I renamed `MyModule` to `Github` and removed the prefixes (i.e., just `Account` and `Organization`) for this reason.

Renaming without removing the prefixes, will get you these types:
- Github
- GithubGitHubOrganization
- GithubGitHubAccount

vikram-dagger replied:
Thanks for this, makes perfect sense. I have made the change for Go and TypeScript in https://github.com/dagger/dagger/pull/6681/commits/d08bafd3ffae799421a7fa8e71cf6ead84eca99e but would like to have your eyes on it @helderco @jedevc @TomChv @vito - thanks!

helderco reviewed (commented):
Just a few notes and comments to explain some of the changes I made in Python.

vikram-dagger commented:
> > > I've udpated the Python docs. I reordered the pages to a flow that makes sense, building from what came before.
> > 
> > 
> > In [874c2d7](https://github.com/dagger/dagger/commit/874c2d7fabc9284813a8d029c9618c01426c6e02) I've moved the dependencies page up in the sequence you suggested, because I think that is an important concept to discuss early. Apart from this, I updated the navbar for the other languages to match your suggestion.
> 
> Yeah, I agree. Also think that adding Python dependencies could be extracted into a page between module structure (I think at one point it was "Project structure", which is more accurate since it's talking mostly about a Python project managed with `pyprojec.toml`), and IDE integration. The latter was also a part of "project structure".

Thanks for this change. I have replicated it in the TypeScript docs as well, hope that's OK @TomChv

vikram-dagger commented on docs/versioned_docs/version-zenith/developer-guide/typescript/406314-aliases.mdx:
```diff
@@ -0,0 +1,28 @@
+
+<PartialExperimentalDocs />
+
+You can change the name of an exposed field or function to make it more readable. To do so, simply add the alias as a parameter of the corresponding decorator.
```
Thanks for the explanation, added to the docs.

vikram-dagger commented on docs/versioned_docs/version-zenith/developer-guide/python/944887-attribute-functions.mdx:
```diff
@@ -0,0 +1,50 @@
+Object attributes can be exposed as Dagger Functions, without having to
+create a getter function explicitly.
+
+There's a [dagger.field][dag-field] descriptor that is a wrapper of
+[dataclasses.field][data-field], but creates a getter function for the attribute
+as well so that it's accessible from the Dagger API.
```
Thanks for this, @helderco . I believe you've already made the necessary changes in the Python docs, but you're suggesting that this could be replicated for the other language guides (please correct me if I'm wrong). In this case and as this is a technical detail that I'm not familiar with, I would prefer to have @jedevc @vito @TomChv to review this comment, also the comments in https://github.com/dagger/dagger/pull/6681#discussion_r1502631807 and https://github.com/dagger/dagger/pull/6681#discussion_r1502629268 and consider whether to make similar changes to the Go and TypeScript docs.

jedevc replied:
I think it makes sense to explain the difference between object field and object function attributes in Go, but the dataclasses construction is python-specific.

vikram-dagger commented on docs/versioned_docs/version-zenith/developer-guide/typescript/817191-custom-types.mdx:
````diff
@@ -0,0 +1,34 @@
+```typescript file=./snippets/custom-types/index.ts
+```
+
+Note that `GitHubOrganization` and `GitHubAccount` have no functions exposed, but these may be defined if needed.
````
Removed, thanks for catching this!
